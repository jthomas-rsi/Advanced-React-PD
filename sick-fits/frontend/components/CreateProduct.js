import { useState } from 'react';
import Router from 'next/router';
import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import useForm from '../lib/useForm';
import Form from './styles/Form';
import DisplayError from './ErrorMessage';
import { ALL_PRODUCTS_QUERY } from './Products';

const CREATE_PRODUCT_MUTATION = gql`
  mutation CREATE_PRODUCT_MUTATION(
    # variables for the mutation
    $name: String!
    $description: String!
    $price: Int!
    $image: Upload
  ) {
    createProduct(
      data: {
        name: $name
        description: $description
        price: $price
        status: "AVAILABLE"
        photo: { create: { image: $image, altText: $name } }
      }
    ) {
      id
      price
      description
      name
    }
  }
`;

const CreateProductComponent = () => {
  const { inputs, handleChange, resetForm, clearForm } = useForm({
    image: '',
    name: 'Cool James',
    price: 12345,
    description: 'This is a really cool guy',
  });

  // utilize useMutation method to run mutation to create a new product via the form
  // deconstruct the return payload from useMutation hook
  const [createProduct, { loading, error, data }] = useMutation(
    CREATE_PRODUCT_MUTATION,
    {
      variables: inputs,
      // set the refetch conditions to refetch the products query after creating a new product
      // this will update the products list with the new product immediately
      refetchQueries: [{ query: ALL_PRODUCTS_QUERY }],
    }
  );

  const handleSubmit = async (e) => {
    e.preventDefault();

    // call the createProduct mutation
    const res = await createProduct();

    // clear the form after creating a new product
    resetForm();

    // once data is created, redirect user to products page
    Router.push({
      pathname: `/product/${res.data.createProduct.id}`,
    });
  };

  return (
    <Form onSubmit={(e) => handleSubmit(e)}>
      <DisplayError error={error} />
      <fieldset disabled={loading} aria-busy={loading}>
        <label htmlFor="image">
          Image
          <input
            required
            type="file"
            id="image"
            name="image"
            onChange={handleChange}
          />
        </label>
        <label htmlFor="name">
          Name
          <input
            type="text"
            id="name"
            name="name"
            placeholder="Name"
            onChange={handleChange}
            value={inputs.name}
          />
        </label>
        <label htmlFor="price">
          Price
          <input
            type="number"
            id="price"
            name="price"
            placeholder="price"
            value={inputs.price}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="description">
          Description
          <textarea
            id="description"
            name="description"
            placeholder="Description"
            onChange={handleChange}
            value={inputs.description}
          />
        </label>
        <button type="submit" onClick={() => {}}>
          + Add Product
        </button>
      </fieldset>
    </Form>
  );
};

export default CreateProductComponent;
