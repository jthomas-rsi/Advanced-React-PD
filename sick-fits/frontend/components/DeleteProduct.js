import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';

const DELETE_PRODUCT_MUTATION = gql`
  mutation DELETE_PRODUCT_MUTATION($id: ID!) {
    deleteProduct(id: $id) {
      id
      name
    }
  }
`;

// function to find and remove product item from the cache
const updateCache = (cache, payload) => {
  // manually update the cache on the client to match the server
  // find the item in the cache and then remove it
  cache.evict(cache.identify(payload.data.deleteProduct));
};

const DeleteProduct = ({ id, children }) => {
  // pass query && id to deleteProduct mutation
  const [deleteProduct, { loading, error }] = useMutation(
    DELETE_PRODUCT_MUTATION,
    {
      variables: { id },
      update: updateCache,
    }
  );

  return (
    <button
      type="button"
      disabled={loading}
      onClick={() => {
        // eslint-disable-next-line no-restricted-globals
        if (confirm('Are you sure you want to delete this item?')) {
          //   console.log('Delete');
          deleteProduct().catch((err) => alert(err.message));
        }
      }}
    >
      {children}
    </button>
  );
};

DeleteProduct.propTypes = {
  id: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
};

export default DeleteProduct;
