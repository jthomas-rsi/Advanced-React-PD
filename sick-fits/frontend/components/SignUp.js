import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import useForm from '../lib/useForm';
import Form from './styles/Form';
import DisplayError from './ErrorMessage';
import { CURRENT_USER_QUERY } from './User';

export const SIGNUP_MUTATION = gql`
  mutation SIGNUP_MUTATION(
    $email: String!
    $name: String!
    $password: String!
  ) {
    createUser(data: { email: $email, name: $name, password: $password }) {
      id
      email
      name
    }
  }
`;

const SignUp = () => {
  // form controls and state
  const { inputs, handleChange, resetForm, clearForm } = useForm({
    name: '',
    email: '',
    password: '',
  });

  // mutation hook
  const [signup, { data, loading, error }] = useMutation(SIGNUP_MUTATION, {
    variables: inputs,
  });

  // handle the form submission
  const handleSubmit = async (e) => {
    e.preventDefault();
    const res = await signup().catch((error) => console.error(error));
    console.log(res);
    console.log({ data, loading, error });
    resetForm();
    // send the email and password to the graphqlAPI
  };

  return (
    <Form method="POST" onSubmit={(e) => handleSubmit(e)}>
      <h2>Sign Up For An Account</h2>
      <DisplayError error={error} />
      <fieldset>
        {data?.createUser && (
          <p>
            Signed up with {data.createUser.email} - Please go ahead and sign
            in!
          </p>
        )}
        <label htmlFor="name">
          Your Name
          <input
            type="text"
            id="name"
            name="name"
            placeholder="Full name"
            autoComplete="name"
            value={inputs.name}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="email">
          Email
          <input
            type="text"
            id="email"
            name="email"
            placeholder="Your email address"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="password">
          Password
          <input
            type="password"
            id="password"
            name="password"
            placeholder="your password"
            value={inputs.password}
            onChange={handleChange}
          />
        </label>
        <button type="submit" onClick={() => {}}>
          Sign Up
        </button>
      </fieldset>
    </Form>
  );
};

export default SignUp;
