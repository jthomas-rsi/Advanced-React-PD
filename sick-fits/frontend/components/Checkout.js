import { loadStripe } from '@stripe/stripe-js';
import { useState } from 'react';
import { useMutation } from '@apollo/client';
import { useRouter } from 'next/router';
import {
  Elements,
  CardElement,
  useStripe,
  useElements,
} from '@stripe/react-stripe-js';
import styled from 'styled-components';
import nProgress from 'nprogress';
import gql from 'graphql-tag';
import { useCart } from '../lib/cartState';
import SickButton from './styles/SickButton';
import CURRENT_USER_QUERY from './User';

const CheckoutFormStyles = styled.form`
  box-shadow: 0 1px 2px 2px rgba(0, 0, 0, 0.4);
  border: 1px solid rgba(0, 0, 0, 0.6);
  border-radius: 5px;
  padding: 1rem;
  display: grid;
  grid-gap: 1rem;
`;

const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER_MUTATION($token: String!) {
    checkout(token: $token) {
      id
      charge
      total
      items {
        id
        name
      }
    }
  }
`;

// TODO replace this string with env variable later
const stripeKey = process.env.NEXT_PUBLIC_STRIPE_KEY;
const stripeLib = loadStripe(stripeKey);

const CheckoutForm = () => {
  const [error, setError] = useState();
  const [loading, setLoading] = useState(false);
  const stripe = useStripe();
  const elements = useElements();
  const router = useRouter();
  const { closeCart } = useCart();
  const [checkout, { error: graphqlError }] = useMutation(
    CREATE_ORDER_MUTATION,
    {
      refetchQueries: [{ query: CURRENT_USER_QUERY }],
    }
  );

  const handleSubmit = async (e) => {
    // 1. stop form from submitting and turn the loader on
    e.preventDefault();
    console.log('We got to do some work...');
    setLoading(true);
    // 2. start the page transition and say "loading"
    nProgress.start();
    // 3. create the payment method via stripe (token comes back here if successful)
    const { error, paymentMethod } = await stripe.createPaymentMethod({
      type: 'card',
      card: elements.getElement(CardElement), // get the card element from the stripe element
    });
    console.log({ paymentMethod });
    // console.log(error);
    // 4. handle any errors from stripe
    if (error) {
      setError(error);
      nProgress.done(); // turn off the progress bar
      return; // stop the checkout flow because there is an error
    }
    // 5. send the token from step 3 to our keystone server via a custom mutation
    const order = await checkout({
      variables: {
        token: paymentMethod.id,
      },
    });

    console.log('Finished with the order!!');
    // 6. change the page to view the order
    console.log({ order });
    router.push({
      pathname: `/order/[id]`,
      query: { id: order.data.checkout.id },
    });
    // 7. close the cart
    closeCart();
    // 8. turn the loader off, set loading to false
    setLoading(false);
    nProgress.done();
  };

  return (
    <CheckoutFormStyles onSubmit={handleSubmit}>
      {error && <p style={{ fontSize: '12px' }}>{error.message}</p>}
      {graphqlError && (
        <p style={{ fontSize: '12px' }}>{graphqlError.message}</p>
      )}
      <CardElement />
      <SickButton>Check Out Now</SickButton>
    </CheckoutFormStyles>
  );
};

const Checkout = () => (
  <Elements stripe={stripeLib}>
    <CheckoutForm />
  </Elements>
);

export default Checkout;
