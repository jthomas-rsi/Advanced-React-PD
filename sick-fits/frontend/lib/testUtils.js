import casual from 'casual';
import { PAGINATION_QUERY } from '../components/Pagination';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  id: '6695cadeb72f324aac2a7f72',
  name: 'KITH Hoodie',
  price: 23562,
  description: 'Love this hoodie',
  photo: {
    id: '6695cadeb72f324aac2a7f71',
    image: {
      publicUrlTransformed:
        'https://res.cloudinary.com/wesbos/image/upload/v1579815935/sick-fits-keystone/5e2a13ff689b2835ae71d1a7.jpg',
      __typename: 'CloudinaryImage_File',
    },
    __typename: 'ProductImage',
  },
  __typename: 'Product',
});

const fakeUser = (overrides) => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
  ...overrides,
});

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  user: fakeUser(),
});

const fakeCartItem = (overrides) => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

function makePaginationMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          _allProductsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: length,
              __typename: 'count',
            },
          },
        },
      },
    },
  ];
}

export {
  makePaginationMocksFor,
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
