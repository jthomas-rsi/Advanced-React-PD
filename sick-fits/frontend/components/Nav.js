import Link from 'next/link';
import NavStyles from './styles/NavStyles';
import { useUser } from './User';
import SignOut from './SignOut';
import { useCart } from '../lib/cartState';
import CartCount from './CartCount';

const Nav = () => {
  const user = useUser();
  const { toggleCart } = useCart();

  const totalItems = user?.cart.reduce(
    (tally, cartItem) => tally + cartItem.quantity,
    0
  );

  // console.log(process.env.NEXT_PUBLIC_STRIPE_PUBLIC_KEY); // variable showing as undefined in console

  return (
    <NavStyles>
      {process.env.NEXT_PUBLIC_STRIPE_PUBLIC_KEY}
      <Link href="/products">Products</Link>
      {user && (
        <>
          <Link href="/sell">Sell</Link>
          <Link href="/orders">Orders</Link>
          <Link href="/account">Account</Link>
          <SignOut />
          <button type="button" onClick={toggleCart}>
            {' '}
            My Cart
            <CartCount count={totalItems} />
          </button>
        </>
      )}
      {!user && <Link href="/signin">Sign In</Link>}
    </NavStyles>
  );
};

export default Nav;
