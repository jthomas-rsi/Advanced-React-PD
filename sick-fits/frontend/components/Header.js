/* eslint-disable react/jsx-props-no-spreading */
import { useState, useEffect } from 'react';
import Link from 'next/link';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Cart from './Cart';
import Nav from './Nav';
import Search from './Search';

const Logo = styled.h1`
  background: var(--red, red);
  font-size: 4rem;
  margin-left: 2rem;
  position: relative;
  z-index: 2;
  transform: skew(-7deg);
  a {
    color: white;
    text-decoration: none;
    text-transform: uppercase;
    padding: 0.5rem 1rem;
  }
`;

const HeaderStyles = styled.header`
  .bar {
    border-bottom: 10px solid var(--black, black);
    display: grid;
    grid-template-columns: auto 1fr;
    justify-content: space-between;
    align-items: stretch;
  }
  .sub-bar {
    display: grid;
    grid-template-columns: 1fr auto;
    border-bottom: 1px solid var(--black, black);
  }
`;

const ClientOnly = ({ children, ...delegated }) => {
  const [hasMounted, setHasMounted] = useState(false);

  useEffect(() => {
    setHasMounted(true);
  }, []);

  if (!hasMounted) {
    return null;
  }

  return <div {...delegated}>{children}</div>;
};

ClientOnly.propTypes = {
  children: PropTypes.node.isRequired,
};

const Header = () => (
  <HeaderStyles>
    <div className="bar">
      <Logo>
        <Link href="/">Sick Fits</Link>
      </Logo>
      <Nav />
    </div>
    <div className="sub-bar">
      <ClientOnly>
        <Search />
      </ClientOnly>
    </div>
    <Cart />
  </HeaderStyles>
);

export default Header;
