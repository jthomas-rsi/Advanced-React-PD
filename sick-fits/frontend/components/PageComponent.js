import PropTypes from 'prop-types';
import Header from './Header';

const PageComponent = ({ children }) => (
  <div>
    <Header />
    <h1>Page Component</h1>
    {children}
  </div>
);

PageComponent.propTypes = {
  children: PropTypes.node.isRequired,
};

export default PageComponent;
