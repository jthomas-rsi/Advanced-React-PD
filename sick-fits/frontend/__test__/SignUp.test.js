import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import SignUp, { SIGNUP_MUTATION } from '../components/SignUp';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser } from '../lib/testUtils';

const mockUser = fakeUser();
const password = 'JamesBond007';
const mocks = [
  // mock mutation
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        email: mockUser.email,
        name: mockUser.name,
        password,
      },
    },
    result: {
      data: {
        createUser: {
          __typename: 'User',
          id: 'abc123',
          email: mockUser.email,
          name: mockUser.name,
        },
      },
    },
  },
];

const TestComponent = () => (
  <MockedProvider mocks={mocks}>
    <SignUp />
  </MockedProvider>
);

describe('SignUp Testing suite ', () => {
  it('Component renders and matches a snapshot', () => {
    const { container } = render(TestComponent());
    expect(container).toMatchSnapshot();
  });

  it('Component calls the mutation properly', async () => {
    const { container, debug } = render(TestComponent());

    // fill in the form
    userEvent.type(screen.getByPlaceholderText(/full name/i), mockUser.name);
    userEvent.type(
      screen.getByPlaceholderText(/Your email address/i),
      mockUser.email
    );
    userEvent.type(screen.getByPlaceholderText(/your password/i), password);

    // submit the form
    userEvent.click(screen.getByRole('button', { name: /sign up/i }));

    // wait for the mutation to complete
    await screen.findByText(/Signed up with/i);

    // There is no need for the expect in this test because if the findByText above fails the entire test will fail
  });
});
