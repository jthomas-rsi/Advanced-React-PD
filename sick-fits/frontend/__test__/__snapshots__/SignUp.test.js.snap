// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp Testing suite  Component renders and matches a snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 jnelna"
    method="POST"
  >
    <h2>
      Sign Up For An Account
    </h2>
    <fieldset>
      <label
        for="name"
      >
        Your Name
        <input
          autocomplete="name"
          id="name"
          name="name"
          placeholder="Full name"
          type="text"
          value=""
        />
      </label>
      <label
        for="email"
      >
        Email
        <input
          id="email"
          name="email"
          placeholder="Your email address"
          type="text"
          value=""
        />
      </label>
      <label
        for="password"
      >
        Password
        <input
          id="password"
          name="password"
          placeholder="your password"
          type="password"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Sign Up
      </button>
    </fieldset>
  </form>
</div>
`;
